script/transform-approved-books-data.ts

```diff
@@ -0,0 +1,168 @@
+}
+
+const isApprovedRepo = (something: any): something is ApprovedRepo => {
+  return typeof something.repo === 'string' && isApprovedBooksData(something);
+};
+
+interface ApprovedVersion {
+  content_version: string;
+  min_code_version: string;
+}
+
+const isApprovedVersion = (something: any): something is ApprovedVersion => {
+  return typeof something.content_version === 'string' && typeof something.min_code_version === 'string';
+};
+
+interface ApprovedVersionCollection extends ApprovedVersion {
+  collection_id: string;
+}
+
+const isApprovedVersionCollection = (something: any): something is ApprovedVersionCollection => {
+  return typeof something.collection_id === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedVersionRepo extends ApprovedVersion {
+  repo: string;
+}
+
+const isApprovedVersionRepo = (something: any): something is ApprovedVersionRepo => {
+  return typeof something.repo === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedBooksAndVersions {
+  approved_books: Array<ApprovedCollection | ApprovedRepo>;
+  approved_versions: Array<ApprovedVersionCollection | ApprovedVersionRepo>;
+}
+
+const isApprovedBooksAndVersions = (something: any): something is ApprovedBooksAndVersions => {
+  return Array.isArray(something.approved_books)
+    && something.approved_books.every(
+      (element: any) => isApprovedCollection(element) || isApprovedRepo(element))
+    && something.approved_versions.every(
+      (element: any) => isApprovedVersionCollection(element) || isApprovedVersionRepo(element));
+};
+
+const matchRepoVersion = (repoData: ApprovedRepo) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionRepo => {
+  return isApprovedVersionRepo(versionData) && versionData.repo === repoData.repo;
+};
+
+const matchCollectionVersion = (collectionData: ApprovedCollection) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionCollection => {
+  return isApprovedVersionCollection(versionData) && versionData.collection_id === collectionData.collection_id;
+};
+
+const getDesiredVersion = (
+  approvedVersions: Array<ApprovedVersionCollection | ApprovedVersionRepo>,
+  colOrRepo: ApprovedRepo | ApprovedCollection
+): string | undefined => {
+  const isCollection = isApprovedCollection(colOrRepo);
+
+  const filter = isCollection
+    ? matchCollectionVersion(colOrRepo as ApprovedCollection)
+    : matchRepoVersion(colOrRepo as ApprovedRepo);
+
+  const transformVersion = (versionData: ApprovedVersionCollection | ApprovedVersionRepo) => {
+    // books from collections have versions starting with 1.
+    return isCollection ? versionData.content_version.slice(2) : versionData.content_version;
+  };
+
+  const versions = approvedVersions
+    // only versions for current repo / collection
+    .filter(filter)
+    .map(transformVersion)
+    // sort and revert so results are descending
+    .sort()
```

Comment:
naive sort doesn't work on semver

Reply:
thats true, however at this step they are already transformed - we are removing first 2 characters so it should be `X.X` for collections and `X` for repos and this should work with `.sort()`

Reply:
yea but like `4.56` is bigger than `4.9` and string or number sort will both get that wrong. you need a semver specific sort

Reply:
right... that's obvious but I didn't know that :grimacing: 
fixed